File: frontend/packages/shared/src/detection/validators.ts
// Validators mirroring aurodlpv2_detection/recognition/validators.py.
// A validator is what separates "twelve digits" from "an Aadhaar number".

const VERHOEFF_D: readonly (readonly number[])[] = [
  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
  [1, 2, 3, 4, 0, 6, 7, 8, 9, 5],
  [2, 3, 4, 0, 1, 7, 8, 9, 5, 6],
  [3, 4, 0, 1, 2, 8, 9, 5, 6, 7],
  [4, 0, 1, 2, 3, 9, 5, 6, 7, 8],
  [5, 9, 8, 7, 6, 0, 4, 3, 2, 1],
  [6, 5, 9, 8, 7, 1, 0, 4, 3, 2],
  [7, 6, 5, 9, 8, 2, 1, 0, 4, 3],
  [8, 7, 6, 5, 9, 3, 2, 1, 0, 4],
  [9, 8, 7, 6, 5, 4, 3, 2, 1, 0],
];

const VERHOEFF_P: readonly (readonly number[])[] = [
  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
  [1, 5, 7, 6, 2, 8, 3, 0, 9, 4],
  [5, 8, 0, 3, 7, 9, 6, 1, 4, 2],
  [8, 9, 1, 6, 0, 4, 3, 5, 2, 7],
  [9, 4, 5, 3, 1, 2, 6, 8, 7, 0],
  [4, 2, 8, 6, 5, 7, 3, 9, 0, 1],
  [2, 7, 9, 3, 8, 0, 6, 4, 1, 5],
  [7, 0, 4, 6, 9, 1, 3, 2, 5, 8],
];

const PAN_HOLDER_TYPES = new Set('PCHABGJLFTEK'.split(''));
const PLACEHOLDER_PANS = new Set([
  'ABCDE1234F',
  'AAAAA0000A',
  'AAAAA1111A',
  'ABCDE0000A',
  'XXXXX0000X',
  'AAAPL1234C',
]);
const GSTIN_ALPHABET = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ';
const STATE_CODES = new Set([
  'AN','AP','AR','AS','BR','CG','CH','DD','DL','DN','GA','GJ','HP','HR','JH','JK',
  'KA','KL','LA','LD','MH','ML','MN','MP','MZ','NL','OD','OR','PB','PY','RJ','SK',
  'TN','TR','TS','UK','UP','WB',
]);

function digitsOnly(value: string): string {
  return value.replace(/\D/g, '');
}

function isRepdigit(digits: string): boolean {
  return new Set(digits.split('')).size === 1;
}

function isSequential(digits: string): boolean {
  let ascending = true;
  let descending = true;
  for (let i = 1; i < digits.length; i++) {
    const delta = Number(digits[i]) - Number(digits[i - 1]);
    if (delta !== 1) ascending = false;
    if (delta !== -1) descending = false;
  }
  return ascending || descending;
}

export function verhoeffOk(value: string): boolean {
  const digits = digitsOnly(value);
  if (!digits) return false;
  let checksum = 0;
  const reversed = digits.split('').reverse();
  for (let i = 0; i < reversed.length; i++) {
    checksum = VERHOEFF_D[checksum]![VERHOEFF_P[i % 8]![Number(reversed[i])]!]!;
  }
  return checksum === 0;
}

export function validateAadhaar(value: string): boolean {
  const digits = digitsOnly(value);
  if (digits.length !== 12) return false;
  if (digits[0] === '0' || digits[0] === '1') return false;
  if (isRepdigit(digits) || isSequential(digits)) return false;
  if (new Set(digits.split('')).size <= 2) return false;
  return verhoeffOk(digits);
}

export function validatePan(value: string): boolean {
  const pan = value.trim().toUpperCase();
  if (!/^[A-Z]{5}\d{4}[A-Z]$/.test(pan)) return false;
  if (PLACEHOLDER_PANS.has(pan)) return false;
  if (!PAN_HOLDER_TYPES.has(pan[3]!)) return false;
  return new Set(pan.slice(0, 5).split('')).size !== 1;
}

export function validateGstin(value: string): boolean {
  const gstin = value.trim().toUpperCase();
  if (!/^[0-3]\d[A-Z]{5}\d{4}[A-Z][1-9A-Z]Z[0-9A-Z]$/.test(gstin)) return false;
  const state = Number(gstin.slice(0, 2));
  if (state < 1 || state > 38) return false;

  let total = 0;
  for (let i = 0; i < 14; i++) {
    const position = GSTIN_ALPHABET.indexOf(gstin[i]!);
    const product = position * (i % 2 ? 2 : 1);
    total += Math.floor(product / 36) + (product % 36);
  }
  return GSTIN_ALPHABET[(36 - (total % 36)) % 36] === gstin[14];
}

export function validateIfsc(value: string): boolean {
  return /^[A-Z]{4}0[A-Z0-9]{6}$/.test(value.trim().toUpperCase());
}

export function validatePassport(value: string): boolean {
  return /^[A-PR-WY]\d{7}$/.test(value.trim().toUpperCase());
}

export function validateDrivingLicense(value: string): boolean {
  const compact = value.trim().toUpperCase().replace(/[\s-]/g, '');
  if (!/^[A-Z]{2}\d{13}$/.test(compact)) return false;
  return STATE_CODES.has(compact.slice(0, 2));
}

export function validateVoterId(value: string): boolean {
  return /^[A-Z]{3}\d{7}$/.test(value.trim().toUpperCase());
}

export function validateInPhone(value: string): boolean {
  let digits = digitsOnly(value);
  if (digits.startsWith('91') && digits.length === 12) digits = digits.slice(2);
  else if (digits.startsWith('0') && digits.length === 11) digits = digits.slice(1);
  if (digits.length !== 10) return false;
  if (!'6789'.includes(digits[0]!)) return false;
  return !isRepdigit(digits);
}

export function validateAbhaNumber(value: string): boolean {
  const digits = digitsOnly(value);
  if (digits.length !== 14) return false;
  if (digits[0] === '0') return false;
  return !isRepdigit(digits);
}

export function validateBankAccount(value: string): boolean {
  const digits = digitsOnly(value);
  if (digits.length < 9 || digits.length > 18) return false;
  return !isRepdigit(digits);
}

export type Validator = (value: string) => boolean;

/**
 * ICD-10 validation is category-level on the client: the full 73,000-code
 * dictionary is too heavy for a content script, but rejecting invalid
 * categories is what stops "room A12" and "vitamin B12" flagging.
 */
export function buildValidators(icd10Categories: readonly string[]): Record<string, Validator> {
  const categories = new Set(icd10Categories);
  return {
    aadhaar: validateAadhaar,
    abha_number: validateAbhaNumber,
    bank_account: validateBankAccount,
    driving_license: validateDrivingLicense,
    gstin: validateGstin,
    icd10: (value) => categories.has(value.trim().toUpperCase().slice(0, 3)),
    ifsc: validateIfsc,
    in_phone: validateInPhone,
    pan: validatePan,
    passport: validatePassport,
    voter_id: validateVoterId,
  };
}

File: detection/aurodlpv2_detection/recognition/validators.py
from __future__ import annotations

import itertools
import re
from collections.abc import Callable
from typing import Final

_VERHOEFF_D: Final[tuple[tuple[int, ...], ...]] = (
    (0, 1, 2, 3, 4, 5, 6, 7, 8, 9),
    (1, 2, 3, 4, 0, 6, 7, 8, 9, 5),
    (2, 3, 4, 0, 1, 7, 8, 9, 5, 6),
    (3, 4, 0, 1, 2, 8, 9, 5, 6, 7),
    (4, 0, 1, 2, 3, 9, 5, 6, 7, 8),
    (5, 9, 8, 7, 6, 0, 4, 3, 2, 1),
    (6, 5, 9, 8, 7, 1, 0, 4, 3, 2),
    (7, 6, 5, 9, 8, 2, 1, 0, 4, 3),
    (8, 7, 6, 5, 9, 3, 2, 1, 0, 4),
    (9, 8, 7, 6, 5, 4, 3, 2, 1, 0),
)

_VERHOEFF_P: Final[tuple[tuple[int, ...], ...]] = (
    (0, 1, 2, 3, 4, 5, 6, 7, 8, 9),
    (1, 5, 7, 6, 2, 8, 3, 0, 9, 4),
    (5, 8, 0, 3, 7, 9, 6, 1, 4, 2),
    (8, 9, 1, 6, 0, 4, 3, 5, 2, 7),
    (9, 4, 5, 3, 1, 2, 6, 8, 7, 0),
    (4, 2, 8, 6, 5, 7, 3, 9, 0, 1),
    (2, 7, 9, 3, 8, 0, 6, 4, 1, 5),
    (7, 0, 4, 6, 9, 1, 3, 2, 5, 8),
)


def _digits(value: str) -> str:
    return re.sub(r"\D", "", value)


def verhoeff_ok(value: str) -> bool:
    digits = _digits(value)
    if not digits:
        return False
    checksum = 0
    for index, char in enumerate(reversed(digits)):
        checksum = _VERHOEFF_D[checksum][_VERHOEFF_P[index % 8][int(char)]]
    return checksum == 0


def _is_repdigit(digits: str) -> bool:
    return len(set(digits)) == 1


def _is_sequential(digits: str) -> bool:
    pairs = list(itertools.pairwise(digits))
    ascending = all(int(b) - int(a) == 1 for a, b in pairs)
    descending = all(int(a) - int(b) == 1 for a, b in pairs)
    return ascending or descending


def validate_aadhaar(value: str) -> bool:
    digits = _digits(value)
    if len(digits) != 12:
        return False
    if digits[0] in {"0", "1"}:
        return False
    if _is_repdigit(digits) or _is_sequential(digits):
        return False
    if len(set(digits)) <= 2:
        return False
    return verhoeff_ok(digits)


_PAN_HOLDER_TYPES: Final[frozenset[str]] = frozenset("PCHABGJLFTEK")
_PAN_RE: Final[re.Pattern[str]] = re.compile(r"^[A-Z]{5}[0-9]{4}[A-Z]$")
_KNOWN_PLACEHOLDER_PANS: Final[frozenset[str]] = frozenset(
    {"ABCDE1234F", "AAAAA0000A", "AAAAA1111A", "ABCDE0000A", "XXXXX0000X", "AAAPL1234C"}
)


def validate_pan(value: str) -> bool:
    pan = value.strip().upper()
    if not _PAN_RE.match(pan):
        return False
    if pan in _KNOWN_PLACEHOLDER_PANS:
        return False
    if pan[3] not in _PAN_HOLDER_TYPES:
        return False
    return len(set(pan[:5])) != 1


_GSTIN_RE: Final[re.Pattern[str]] = re.compile(
    r"^[0-3][0-9][A-Z]{5}[0-9]{4}[A-Z][1-9A-Z]Z[0-9A-Z]$"
)
_GSTIN_ALPHABET: Final[str] = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"


def validate_gstin(value: str) -> bool:
    gstin = value.strip().upper()
    if not _GSTIN_RE.match(gstin):
        return False
    state = int(gstin[:2])
    if not 1 <= state <= 38:
        return False

    total = 0
    for index, char in enumerate(gstin[:14]):
        position = _GSTIN_ALPHABET.index(char)
        factor = 2 if index % 2 else 1
        product = position * factor
        total += product // 36 + product % 36
    check = (36 - total % 36) % 36
    return _GSTIN_ALPHABET[check] == gstin[14]


_IFSC_RE: Final[re.Pattern[str]] = re.compile(r"^[A-Z]{4}0[A-Z0-9]{6}$")


def validate_ifsc(value: str) -> bool:
    return bool(_IFSC_RE.match(value.strip().upper()))


_PASSPORT_RE: Final[re.Pattern[str]] = re.compile(r"^[A-PR-WY][0-9]{7}$")


def validate_passport(value: str) -> bool:
    return bool(_PASSPORT_RE.match(value.strip().upper()))


_DL_RE: Final[re.Pattern[str]] = re.compile(r"^[A-Z]{2}[\s-]?[0-9]{2}[\s-]?[0-9]{4}[0-9]{7}$")
_STATE_CODES: Final[frozenset[str]] = frozenset(
    {
        "AN",
        "AP",
        "AR",
        "AS",
        "BR",
        "CG",
        "CH",
        "DD",
        "DL",
        "DN",
        "GA",
        "GJ",
        "HP",
        "HR",
        "JH",
        "JK",
        "KA",
        "KL",
        "LA",
        "LD",
        "MH",
        "ML",
        "MN",
        "MP",
        "MZ",
        "NL",
        "OD",
        "OR",
        "PB",
        "PY",
        "RJ",
        "SK",
        "TN",
        "TR",
        "TS",
        "UK",
        "UP",
        "WB",
    }
)


def validate_driving_license(value: str) -> bool:
    compact = re.sub(r"[\s-]", "", value.strip().upper())
    if not re.match(r"^[A-Z]{2}[0-9]{13}$", compact):
        return False
    return compact[:2] in _STATE_CODES


_VOTER_RE: Final[re.Pattern[str]] = re.compile(r"^[A-Z]{3}[0-9]{7}$")


def validate_voter_id(value: str) -> bool:
    return bool(_VOTER_RE.match(value.strip().upper()))


def validate_icd10(value: str) -> bool:
    try:
        import simple_icd_10_cm as icd
    except ImportError:  # pragma: no cover - dependency is declared
        return False
    return bool(icd.is_valid_item(value.strip().upper()))


def validate_in_phone(value: str) -> bool:
    digits = _digits(value)
    if digits.startswith("91") and len(digits) == 12:
        digits = digits[2:]
    elif digits.startswith("0") and len(digits) == 11:
        digits = digits[1:]
    if len(digits) != 10:
        return False
    if digits[0] not in {"6", "7", "8", "9"}:
        return False
    return not _is_repdigit(digits)


def validate_abha_number(value: str) -> bool:
    digits = _digits(value)
    if len(digits) != 14:
        return False
    if digits[0] == "0":
        return False
    return not _is_repdigit(digits)


def validate_bank_account(value: str) -> bool:
    digits = _digits(value)
    if not 9 <= len(digits) <= 18:
        return False
    return not _is_repdigit(digits)


def always_valid(value: str) -> bool:
    del value
    return True


VALIDATORS: Final[dict[str, Callable[[str], bool]]] = {
    "aadhaar": validate_aadhaar,
    "abha_number": validate_abha_number,
    "bank_account": validate_bank_account,
    "driving_license": validate_driving_license,
    "gstin": validate_gstin,
    "icd10": validate_icd10,
    "ifsc": validate_ifsc,
    "in_phone": validate_in_phone,
    "pan": validate_pan,
    "passport": validate_passport,
    "voter_id": validate_voter_id,
}


def get_validator(name: str | None) -> Callable[[str], bool]:
    if name is None:
        return always_valid
    validator = VALIDATORS.get(name)
    if validator is None:
        raise KeyError(f"unknown validator: {name}")
    return validator
